src/Database/Query.php

```diff
@@ -1328,6 +1330,29 @@ public function andHaving($conditions, $types = [])
         return $this;
     }
 
+    /**
+     * Adds a named window expression.
+     *
+     * If the window exprssion is copying from an existing window name, the
+     * target window expression must be added to the query first.
+     *
+     * @param string $name Window name
+     * @param \Cake\Database\Expression\WindowExpression $window Window expression
+     * @param bool $overwrite Clear all previous query window expressions
+     * @return $this
+     */
+    public function window(string $name, WindowExpression $window, bool $overwrite = false)
```

Comment:
Any chance for closure support for `$window`, similar to the other query builder methods, so we could do stuff like:

```php
->window('win', function (Query $query) {
    return (new WindowExpression())
        ->partition($query->func()->foo([$query->identifier('bar')])); 
})
```

It would also give me one more reason to ask for expression support for frame bound offsets 😉

Reply:
I can add closure support, but as your example shows we'll have to break with the convention of passing a `QueryExpression`.

Reply:
Maybe we should add `WindowExpression::setName()` and pass the function a `new WindowExpression()` instance similar to how other builders pass an expression and expect it in return.

I was avoiding `->setName('other_window')->partition()` because I thought it looked odd.

Reply:
Oh I just omitted that by accident :) But I also wouldn't have a problem with it not being passed, one can always easily do `$query->newExpr()`.

Reply:
But `QueryExpression` has no use here. We'd need to pass a `WindowExpression` instance which they could return.

Reply:
You mean like this?

```php
->window(function (WindowExpression $window) {
    return $window->setName('win') ...
})
```

I guess that could work, it's similar enough to how you get expression objects in other methods.

Reply:
That's what I'm implementing. You're free to construct your own expression of course, but want a consistent experience.

Reply:
Ok, now I have to think about whether that might make sense for CTEs too :)

Reply:
That's true -- but CTE is awkward because the name is required, but I suppose you can create an invalid expression in other builders that accept a closure.

Reply:
This is the wrong place -- but while I'm thinking about it, we can just have: `->with('name', $cte)` which accepts either closure or expression. That would solve the awkwardness.